### ISSUE TITLE
One of the Ghacks Settings Breaks DoH

### ISSUE THREAD

**Issue by travankor** [ghacksuserjs/ghacks-user.js#835]
I want to use a custom DoH server with Firefox's TRR (in other words no Cloudflare).

I add the following to the user.js:
`network.trr.bootstrapAddress`: IP address of DoH Server
`network.trr.mode`: 3
`network.trr.uri`: DoH server URI
`network.security.esni.enabled`: true

These DoH  settings work on a clean Firefox with no custom user.js (verified with https://dnsleaktest.com). However, neither Ghacks nor pyllyukko's user.js works. I think one of the custom settings is breaking the DoH, but I don't know which ones to look at.

Version: FF 70 Linux (Wayland Version)

**Comment by earthlng** [comment]
Any relevant error messages in the console?
Can you share the actual IP and URL of the DoH server you're trying to use?

**Comment by travankor** [comment]
Here's the console output:
```
JavaScript error: resource:///modules/sessionstore/SessionStore.jsm, line 1356: uncaught exception: 2147746065
JavaScript error: resource://gre/modules/XULStore.jsm, line 66: Error: Can't find profile directory.
JavaScript error: resource://gre/modules/XULStore.jsm, line 66: Error: Can't find profile directory.
```

For every new website I try to reach, I get another of the `line 66: Can't find profile directory` error.

Here's a list of public DoH servers: https://github.com/curl/curl/wiki/DNS-over-HTTPS

From that list I tried, these three:
https://dns.adguard.com/dns-query
https://doh.securedns.eu/dns-query
https://doh.cleanbrowsing.org/doh/family-filter/

I use the `drill` tool from NLnet's ldns to find their IP address for `bootstrapAddress`.

**Comment by Thorin-Oakenpants** [comment]
may or may not be related (this error message seems to be slightly common)
- https://bugzilla.mozilla.org/show_bug.cgi?id=1587439 <-- one that was fixed in 71
- https://bugzilla.mozilla.org/show_bug.cgi?id=1577180 <-- one that was fixed in 70

Does it happen in Nightly, or Dev?

**Comment by travankor** [comment]
I can reproduce the error with Nightly (FF 72.0a1), ie: DNS works on a clean config but not on ghack's config. I should mention that Cloudflare DNS works.

```
[Exception... "Component returned failure code: 0x80520012 (NS_ERROR_FILE_NOT_FOUND) [nsIXPCComponents_Utils.readUTF8URI]"  nsresult: "0x80520012 (NS_ERROR_FILE_NOT_FOUND)"  location: "JS frame :: resource://gre/modules/L10nRegistry.jsm :: this.L10nRegistry.loadSync :: line 759"  data: no] 6 L10nRegistry.jsm:759:19
[fluent] Missing translations in en-CA: appmenuitem-zoom-reduce, appmenuitem-zoom-enlarge. Localization.jsm:202:13
[Exception... "Component returned failure code: 0x80004001 (NS_ERROR_NOT_IMPLEMENTED) [nsIAppStartup.secondsSinceLastOSRestart]"  nsresult: "0x80004001 (NS_ERROR_NOT_IMPLEMENTED)"  location: "JS frame :: resource:///modules/BrowserGlue.jsm :: _collectStartupConditionsTelemetry :: line 1550"  data: no] BrowserGlue.jsm:1550:9
    _collectStartupConditionsTelemetry resource:///modules/BrowserGlue.jsm:1550
    BG__onFirstWindowLoaded resource:///modules/BrowserGlue.jsm:1649
    BG_observe resource:///modules/BrowserGlue.jsm:871
    _delayedStartup chrome://browser/content/browser.js:2099
    _delayedStartup self-hosted:868
Unknown category for SetEventRecordingEnabled: fxmonitor
1573005200647	app.normandy.bootstrap	ERROR	Failed to initialize about pages:: TypeError: AboutPages.init is not a function(resource://normandy/Normandy.jsm:95:24) JS Stack trace: finishInit@Normandy.jsm:95:24
async*observe@Normandy.jsm:71:12
initializeWindow@SessionStore.jsm:1356:22
onBeforeBrowserWindowShown/<@SessionStore.jsm:1534:16
promise callback*onBeforeBrowserWindowShown@SessionStore.jsm:1517:8
ssi_observe@SessionStore.jsm:860:14
onLoad@browser.js:1844:18
EventHandlerNonNull*@browser.xhtml:121:39 3 Log.jsm:723
    append resource://gre/modules/Log.jsm:723
    log resource://gre/modules/Log.jsm:379
    error resource://gre/modules/Log.jsm:387
    finishInit resource://normandy/Normandy.jsm:97
    InterpretGeneratorResume self-hosted:1149
    AsyncFunctionNext self-hosted:688
Could not parse ASRouter preference. Try resetting browser.newtabpage.activity-stream.asrouter.providers.snippets in about:config. ASRouterPreferences.jsm:61
    _getProviderConfig resource://activity-stream/lib/ASRouterPreferences.jsm:61
    reduce self-hosted:295
    _getProviderConfig resource://activity-stream/lib/ASRouterPreferences.jsm:56
    get providers resource://activity-stream/lib/ASRouterPreferences.jsm:74
    _updateMessageProviders resource://activity-stream/lib/ASRouter.jsm:593
    init resource://activity-stream/lib/ASRouter.jsm:841
    InterpretGeneratorResume self-hosted:1149
    AsyncFunctionNext self-hosted:688
uncaught exception: 2147746065
Key event not available on some keyboard layouts: key=“r” modifiers=“accel,alt” id=“key_toggleReaderMode” browser.xhtml
Key event not available on some keyboard layouts: key=“i” modifiers=“accel,alt,shift” id=“key_browserToolbox” browser.xhtml
1573005259838	Telemetry::CoveragePing	ERROR	no endpoint base set Log.jsm:723
LoginHelper: Couldn't parse origin for about:config 
Exception
​
columnNumber: 0
​
data: null
​
filename: "resource://gre/modules/LoginHelper.jsm"
​
lineNumber: 244
​
location: XPCWrappedNative_NoHelper { QueryInterface: QueryInterface(), filename: Getter, name: Getter, … }
​
message: "Component returned failure code: 0x80004005 (NS_ERROR_FAILURE) [nsIURI.displayHostPort]"
​
name: "NS_ERROR_FAILURE"
​
result: 2147500037
​
stack: "getLoginOrigin@resource://gre/modules/LoginHelper.jsm:244:7\ngetFormActionOrigin@resource://gre/modules/LoginHelper.jsm:264:17\ncreateFromForm@resource://gre/modules/LoginFormFactory.jsm:62:35\ncreateFromField@resource://gre/modules/LoginFormFactory.jsm:107:19\ngetUserNameAndPasswordFields@resource://gre/modules/LoginManagerChild.jsm:2198:33\ngetFieldContext@resource://gre/modules/LoginManagerChild.jsm:2234:50\nhandleEvent@resource:///actors/ContextMenuChild.jsm:570:7\n"
​
<prototype>: ExceptionPrototype { toString: toString(), name: Getter, message: Getter, … }
LoginHelper.jsm:248:11
LoginHelper: Couldn't parse origin for about:config 
Exception
​
columnNumber: 0
​
data: null
​
filename: "resource://gre/modules/LoginHelper.jsm"
​
lineNumber: 244
​
location: XPCWrappedNative_NoHelper { QueryInterface: QueryInterface(), filename: Getter, name: Getter, … }
​
message: "Component returned failure code: 0x80004005 (NS_ERROR_FAILURE) [nsIURI.displayHostPort]"
​
name: "NS_ERROR_FAILURE"
​
result: 2147500037
​
stack: "getLoginOrigin@resource://gre/modules/LoginHelper.jsm:244:7\ngetUserNameAndPasswordFields@resource://gre/modules/LoginManagerChild.jsm:2201:34\ngetFieldContext@resource://gre/modules/LoginManagerChild.jsm:2234:50\nhandleEvent@resource:///actors/ContextMenuChild.jsm:570:7\n"
​
<prototype>: ExceptionPrototype { toString: toString(), name: Getter, message: Getter, … }
LoginHelper.jsm:248:11
LoginRecipes: getRecipes: falling back to a synchronous message for: null LoginRecipes.jsm:265:9
TypeError: this.contentWindow.windowRoot is null PictureInPictureChild.jsm:308:5
TypeError: "NetworkError when attempting to fetch resource." PublicSuffixList.jsm:34:29
    init resource://gre/modules/netwerk-dns/PublicSuffixList.jsm:34
getScreenshot(http://startpage.com/) failed: TypeError: NetworkError when attempting to fetch resource. Screenshots.jsm:59
    getScreenshotForURL resource://activity-stream/lib/Screenshots.jsm:59
    AsyncFunctionThrow self-hosted:692
getScreenshot(https://dnsleaktest.com/) failed: TypeError: NetworkError when attempting to fetch resource. Screenshots.jsm:59
    getScreenshotForURL resource://activity-stream/lib/Screenshots.jsm:59
    AsyncFunctionThrow self-hosted:692
```

**Comment by Thorin-Oakenpants** [comment]
OK, so if you have time, use the halving method in the [troubleshooting wiki](https://github.com/ghacksuserjs/ghacks-user.js/wiki/1.4-Troubleshooting)

tl;dr the wiki
 - Set up a new profile with your DoH setting, close, and copy it as master
 - apply **half** the user.js and test
    - if it still works add the **next quarter** and restart and re-test
       - if it still works add the **next eighth** and restart and re-test
   - if the test fails, close and reset your profile to the master done in step one
      - I think you get where this is going

**Comment by travankor** [comment]
Thanks, I will see if I can do this over the weekend. But this approach basically boils down to manually testing which option breaks my setup with trial and error.

So no idea or guesses where to start?

Note: I am using the vanilla user.js, with all the default comments and all. (I used to use pyllyukko's user.js in the past but he seems to be understandably less active these days...)

**Comment by Thorin-Oakenpants** [comment]
> But this approach basically boils down to manually testing which option breaks my setup with trial and error

Yes, but it's a **binary** method, so it very quickly narrows it down. When I've used it, it usually only takes three or four tests to find the section, and that instantly leads to ideas to flip in about:config

The only way to find the culprit is to test: there's no magic bullet

**Comment by earthlng** [comment]
Hi again,

I tested it with a new profile, the latest master user.js and using the adguard DoH:
```js
user_pref("network.trr.bootstrapAddress", "176.103.130.130"); // found @ https://kb.adguard.com/en/dns/setup-guide
user_pref("network.trr.mode", 3);
user_pref("network.trr.uri", "https://dns.adguard.com/dns-query");
user_pref("network.security.esni.enabled", true);
```
and it works just fine. If it doesn't work for you, I'd start narrowing it down by leaving the bootstrap address empty and see if that works.

The "Can't find profile directory" error messages seem unrelated and the screenshot ones are caused by us disabling the background pagethumbs thingy with `2604` but that just prevents the screenshots sysaddon from capturing a thumbnail which is ultimately unrelated to DoH and taking an actual screenshot still works too. You can either enable 2604 or disable the screenshots addon if those messages bother you.

**Comment by earthlng** [comment]
To get a very detailed TRR log go to `about:networking#logging`, set the Log Modules to `nsHostResolver:5` and start logging to a file of your choice. That should help pinpoint what the problem is

**Comment by travankor** [comment]
I have the following for nsHostResolver:5:
```
[(null) 8145: Main Thread]: D/nsHostResolver TRR::Observe() topic=nsPref:changed
[(null) 8145: Main Thread]: D/nsHostResolver TRR::Observe() topic=nsPref:changed
[(null) 8145: Main Thread]: D/nsHostResolver TRR::Observe() topic=nsPref:changed
[(null) 8145: Main Thread]: D/nsHostResolver TRRService starting confirmation test https://mozilla.cloudflare-dns.com/dns-query example.com
[(null) 8145: Main Thread]: D/nsHostResolver TRR::Observe() topic=nsPref:changed
[(null) 8145: Main Thread]: D/nsHostResolver TRRService clearing blacklist because of change is uri service
[(null) 8145: Main Thread]: D/nsHostResolver TRR::SendHTTPRequest resolve example.com type 2
[(null) 8145: Socket Thread]: D/nsHostResolver TRRService::MaybeBootstrap: use 176.103.130.130 instead of dns.adguard.com
[(null) 8145: Socket Thread]: D/nsHostResolver Resolving host [176.103.130.130] type 0. [this=0x7f9d4d27b480]
[(null) 8145: Socket Thread]: D/nsHostResolver   Host is IP Literal [176.103.130.130].
[(null) 8145: Socket Thread]: D/nsHostResolver Resolving host [_esni.dns.adguard.com] type 16. [this=0x7f9d4d27b480]
[(null) 8145: Socket Thread]: D/nsHostResolver   No usable record in cache for host [_esni.dns.adguard.com] type 16.
```

```
[Parent 8425: Socket Thread]: D/nsHostResolver Resolving host [wikipedia.org] type 0. [this=0x7fb72237b480]
[Parent 8425: Socket Thread]: D/nsHostResolver   No usable record in cache for host [wikipedia.org] type 0.
[Parent 8425: Socket Thread]: D/nsHostResolver TRRService::Enabled mConfirmationState=3 mCaptiveIsPassed=0
[Parent 8425: Socket Thread]: D/nsHostResolver TrrLookup:: wikipedia.org service not enabled
[Parent 8425: Socket Thread]: D/nsHostResolver Resolving host [dnsleaktest.com] type 0. [this=0x7fb72237b480]
[Parent 8425: Socket Thread]: D/nsHostResolver   No usable record in cache for host [dnsleaktest.com] type 0.
[Parent 8425: Socket Thread]: D/nsHostResolver TRRService::Enabled mConfirmationState=3 mCaptiveIsPassed=0
[Parent 8425: Socket Thread]: D/nsHostResolver TrrLookup:: dnsleaktest.com service not enabled
[Parent 8425: Main Thread]: D/nsHostResolver TRR::Observe() topic=browser:purge-session-history
[Parent 8425: Main Thread]: D/nsHostResolver Shutting down host resolver.
```

I use the same settings as you did. [Longer log](https://paste.debian.net/1115313/) where I messed around with `bootstrapAddress`.

**Comment by travankor** [comment]
I re-installed firefox to see if that would fix this, but now I have more problems O.o

Instead of using the entire window to store the webpage, there is this [grey box ](https://files.catbox.moe/yghxxy.png) around the webpage. I can't interact with the box, nor can I remove it. Note that this is only happening with Ghack's user.js, not with a clean one. I think this might be an OS-related issue (Firefox Linux Wayland).

```
[Exception... "Component returned failure code: 0x80004001 (NS_ERROR_NOT_IMPLEMENTED) [nsIAppStartup.secondsSinceLastOSRestart]"  nsresult: "0x80004001 (NS_ERROR_NOT_IMPLEMENTED)"  location: "JS frame :: resource:///modules/BrowserGlue.jsm :: _collectStartupConditionsTelemetry :: line 1540"  data: no] BrowserGlue.jsm:1540:9
uncaught exception: 2147746065 SessionStore.jsm:1356:22
Unknown category for SetEventRecordingEnabled: fxmonitor
Could not parse ASRouter preference. Try resetting browser.newtabpage.activity-stream.asrouter.providers.snippets in about:config. ASRouterPreferences.jsm:61
Key event not available on some keyboard layouts: key=“r” modifiers=“accel,alt” id=“key_toggleReaderMode” browser.xhtml
Key event not available on some keyboard layouts: key=“r” modifiers=“accel,alt” id=“key_quickRestart” browser.xhtml
Key event not available on some keyboard layouts: key=“i” modifiers=“accel,alt,shift” id=“key_browserToolbox” browser.xhtml
Error: Can't find profile directory. XULStore.jsm:66:15
1573255405565	Telemetry::CoveragePing	ERROR	no endpoint base set Log.jsm:723
getScreenshot(http://startpage.com/) failed: TypeError: NetworkError when attempting to fetch resource. Screenshots.jsm:59
Use of nsIFile in content process is deprecated. 7 NetUtil.jsm:253:8
NS_ERROR_FILE_ACCESS_DENIED: Failed to open input source 'file:///' DevToolsUtils.js:678
TypeError: this.textbox is undefined 27 UrlbarInput.jsm:133:9
Error: Can't find profile directory. XULStore.jsm:66:15
Content Security Policy: Ignoring ‘x-frame-options’ because of ‘frame-ancestors’ directive.
TypeError: this.textbox is undefined 21 UrlbarInput.jsm:133:9
getScreenshot(https://www.startpage.com/assets/images/dtc/1539093424-previewfb.png) failed: TypeError: NetworkError when attempting to fetch resource. Screenshots.jsm:59
TypeError: this.textbox is undefined UrlbarInput.jsm:133:9
TypeError: this.textbox is undefined 2 UrlbarInput.jsm:133:9
```

**Comment by earthlng** [comment]
the grey box is probably the letterboxing, see pref `4504`.

as for your TRR problem I found this [1556194](https://bugzilla.mozilla.org/show_bug.cgi?id=1556194) which talks about OCSP problems with TRR mode 3.
Try setting `security.OCSP.require` to false and if that doesn't work, temporarily disable OCSP with `security.OCSP.enabled=0` to rule out that the problem is OCSP-related

There are also some issues with IPv6: https://bugzilla.mozilla.org/show_bug.cgi?id=1589781

and apparently you can't use mode 3 without a bootstrap address, so scratch that, sorry.
see https://bugzilla.mozilla.org/show_bug.cgi?id=1561084

If none of that works, have you tried using our troubleshooter scratchpad? It contains most of the problematic prefs from our user.js and there's a good chance it can pinpoint the culprit

**Comment by travankor** [comment]
>the grey box is probably the letterboxing, see pref 4504.

Oh ok, I was scared I had broken something, lol.

>So, it looks like the TLS handshake is never completed or completed too late after 1500ms.
I also noticed that every OCSP request is failed because of NS_ERROR_UNKNOWN_HOST.
Maybe it's a deadlock that TLS handshake is waiting for OCSP requets and OCSP request is waiting for the TRR channel to be ready.

 [1556194](https://bugzilla.mozilla.org/show_bug.cgi?id=1556194) seems to be the issue. `security.OCSP.require` to false seems to fix it. Cloudflare works regardless of what `security.OCSP.require` says.

>If none of that works, have you tried using our troubleshooter scratchpad? It contains most of the problematic prefs from our user.js and there's a good chance it can pinpoint the culprit

that probably wouldn't have worked, since the TRR only needs you to set security.OCSP.require to false only at startup or when you sleep/hibernate (otherwise it works with `security.OCSP.require` set to true). so whitelisting all prefs at the beginning wouldn't detect anything.

So thank you guys on this issue. Honestly, I probably wouldn't use firefox if you guys weren't around. Cheers. 🍻

**Comment by earthlng** [comment]
you're welcome. Cheers 🍻
